tanjun/clients.py

```diff
@@ -2697,29 +2689,33 @@ def _reload_module(
         _LOGGER.info("Reloading %s", module_path)
 
         old_loaders = _get_loaders(old_module, module_path)
-        # We assert that the old module has unloaders early to avoid unnecessarily
-        # importing the new module.
-        if not any(loader.has_unload for loader in old_loaders):
-            raise errors.ModuleMissingUnloaders(f"Didn't find any unloaders in old {module_path}", module_path)
+        with _WrapLoadError(errors.FailedModuleUnload, module_path):
+            self._call_unloaders(module_path, old_loaders)
 
         module = yield load_module
 
-        loaders = _get_loaders(module, module_path)
+        if not module:
+            # Indicates the newer version of this module couldn't be imported
+            try:
+                self._call_loaders(module_path, old_loaders)
 
-        # We assert that the new module has loaders early to avoid unnecessarily
-        # unloading then rolling back when we know it's going to fail to load.
-        if not any(loader.has_load for loader in loaders):
-            raise errors.ModuleMissingLoaders(f"Didn't find any loaders in new {module_path}", module_path)
+            except Exception as exc:
+                # TODO: exc here is annoyingly already chained with another
+                # error which would've already been logged
+                _LOGGER.debug("Failed to revert %s", module_path, exc_info=exc)
+                modules_dict.pop(module_path)
 
-        with _WrapLoadError(errors.FailedModuleUnload, module_path):
-            # This will never raise MissingLoaders as we assert this earlier
-            self._call_unloaders(module_path, old_loaders)
+            return
+
+        loaders = _get_loaders(module, module_path)
 
         try:
-            # This will never raise MissingLoaders as we assert this earlier
             self._call_loaders(module_path, loaders)
+        except errors.ModuleMissingLoaders:
+            modules_dict.pop(module_path)
+            raise
         except Exception as exc:
-            self._call_loaders(module_path, old_loaders)
+            modules_dict.pop(module_path)
```

Comment:
failed module load could have a variant with a field which indicates whether it reverted or not